core/core.pro: Assure that libsensordatatypes-qt5.so has already been…

sunweaver opened the pull request:
… built before linking libsensorfw-qt5.so against it.

pvuorela commented on core/core.pro:
```diff
@@ -17,8 +17,7 @@ INCLUDEPATH += $$SENSORFW_INCLUDEPATHS
 
-QMAKE_LIBDIR_FLAGS += -L../datatypes\
-                      -lsensordatatypes-qt5
+LIBS += -L$$[QT_INSTALL_LIBS] -L../datatypes -lsensordatatypes-qt5
```
This doesn't look right. The datatypes should be linked against the built version, not the installed one. And the root .pro file anyway has "CONFIG += ordered" and datatypes before the core so the library should be there?

sunweaver replied:
We (UBports) definitely had race-condition-like build failures without this patch from time to time.

This patch basically mimics what is done in
https://github.com/sailfishos/sensorfw/blob/master/core/hybris.pro#L18

sunweaver replied:
See https://ci.ubports.com/blue/organizations/jenkins/UBportsCore%2FPackaging%20%252F%20sensorfw/detail/MR-12/2/pipeline

pvuorela replied:
On hybris the current form with -L$$[QT_etc coming from commit 1dd09c7b91c. Earlier it was linking against hybris libs so it made some sense, but on that commit I suppose it might be a leftover and the local directories could suffice.

cc @mlehtima ^

Regardless I'm not following how this change would fix your build since the LIBS shouldn't affect the order of build nor dependencies. Maybe works by accident if using an earlier installed library from the system(?) 

Also puzzled why the ordered build is not honored on your build. But if it helps, it could be nice to have that replaced with subdirectory dependencies a la https://wiki.qt.io/SUBDIRS_-_handling_dependencies

sunweaver replied:
On our build machines the libsensorfw-qt5.so.0 shall be linked against -lsensordatatypes-qt5, but the libsensordatatypes-qt5.so symlink has not been created by then, so the to-be-linked against library is not found, see:

```
rm -f libsensordatatypes-qt5.so.1.0.0 libsensordatatypes-qt5.so libsensordatatypes-qt5.so.1 libsensordatatypes-qt5.so.1.0
g++ -Wl,-Bsymbolic-functions -Wl,-z,relro -shared -Wl,-soname,libsensordatatypes-qt5.so.1 -o libsensordatatypes-qt5.so.1.0.0 xyz.o orientation.o unsigned.o compass.o utils.o tap.o lid.o moc_xyz.o moc_orientation.o moc_unsigned.o moc_magneticfield.o moc_compass.o moc_datarange.o moc_tap.o moc_proximity.o moc_lid.o  -lhardware -lpthread /usr/lib/x86_64-linux-gnu/libQt5DBus.so /usr/lib/x86_64-linux-gnu/libQt5Core.so    
[...]
g++ -Wl,-Bsymbolic-functions -Wl,-z,relro -shared -Wl,-soname,libsensorfw-qt5.so.0 -o libsensorfw-qt5.so.0.9.0 sensormanager.o sensormanager_a.o pusher.o ringbuffer.o producer.o source.o consumer.o bin.o filter.o deviceadaptor.o loader.o plugin.o abstractsensor_a.o abstractsensor.o parameterparser.o abstractchain.o sysfsadaptor.o sockethandler.o inputdevadaptor.o config.o nodebase.o moc_sensormanager.o moc_sensormanager_a.o moc_deviceadaptor.o moc_plugin.o moc_abstractsensor_a.o moc_abstractsensor.o moc_abstractchain.o moc_sysfsadaptor.o moc_sockethandler.o moc_nodebase.o  -L../datatypes -lsensordatatypes-qt5 -lhardware -lpthread /usr/lib/x86_64-linux-gnu/libQt5Network.so /usr/lib/x86_64-linux-gnu/libQt5DBus.so /usr/lib/x86_64-linux-gnu/libQt5Core.so    
ln -s libsensordatatypes-qt5.so.1.0.0 libsensordatatypes-qt5.so
ln -s libsensordatatypes-qt5.so.1.0.0 libsensordatatypes-qt5.so.1
ln -s libsensordatatypes-qt5.so.1.0.0 libsensordatatypes-qt5.so.1.0
[...]
```

With my proposed patch, the libsensorfw-qt5.so.0 linking (its entire build, to be exact) waits until libsensordatatypes-qt5.so is fully built. So, the LIBS line declares a dependency between those targets.

See: https://stackoverflow.com/questions/37866187/libs-vs-pre-targetdeps-in-pro-files-of-qt

sunweaver replied:
This discussion is obsolete as the patch has entirely changed.

peat-psuwit committed:
sensorfw.pro: fix linking race condition by avoiding duplicated SUBDIRS

Having duplicated entry create duplicated rules in Makefile, which
seems to confuse Makefile's dependency ordering. To achived intended
effect, remove the old entry before adding the new ones.

sunweaver commented:
We looked into this a bit deeper and @peat-psuwit came up with a completely different solution addressing the same problem (irregular FTBFSes).

mlehtima commented on sensorfw.pro:
```diff
@@ -64,6 +64,8 @@ contains(CONFIG,hybris) {
 } else {
     config_hybris {
+    # Reorder so that adaptors are built after hybris.
+    SUBDIRS -= adaptors
```
Would it be cleaner to move the SUBDIRS += adaptors from near the beginning of the file to the end of the file or at least below this contains-else part and remove adding adaptors from both here and above in the contains part of the code?

spiiroin replied:
AFAICT this works / makes difference because there is "CONFIG += ordered" present -> fixes an issue -> already useful as it is.

But given that sensorfwd builds are painfully slow I started to wonder if it would be possible to drop that CONFIG=ordered bit, use explicit dependencies ("core/hybris.pro.depends += adaptors" or however it is done when a SUBDIRS element is not a plain subdirectory?) and thus enable parallel builds? And whether that would make any difference from build time pov?

rainemak commented:
Thanks @sunweaver ! By any chance would you be able to rebase your PR?

sunweaver commented:
> Thanks @sunweaver ! By any chance would you be able to rebase your PR?

rebased
